project/src/main/java/com/google/ensemble/servlets/UploadServlet.java

```diff
@@ -0,0 +1,101 @@
+// limitations under the License.
+
+package com.google.ensemble.servlets;
+
+import com.google.api.core.ApiFuture;
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.cloud.firestore.Firestore;
+import com.google.cloud.firestore.FirestoreOptions;
+import com.google.cloud.firestore.WriteResult;
+import com.google.ensemble.data.Constants;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import javax.annotation.Nullable;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.lang.InterruptedException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * Servlet that stores and returns visitor comments
+ */
+@WebServlet(Constants.SERVLET_UPLOAD)
+public class UploadServlet extends HttpServlet {
+
+  private static final String FILE_INPUT_NAME = "file";
+  private static final BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+  private static final FirestoreOptions firestoreOptions =
+    FirestoreOptions.getDefaultInstance().toBuilder()
+    .setProjectId(Constants.PROJECT_ID)
+    .build();
+  private static final Firestore db = firestoreOptions.getService();
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String project = request.getParameter("project");
+    String filename = request.getParameter("filename");
+    String fileUrl = getUploadedFileUrl(request, FILE_INPUT_NAME);
+
+    Map<String, Object> docData = new HashMap<>();
+    docData.put("filename", filename);
+    docData.put("fileUrl", fileUrl);
+    ApiFuture<WriteResult> future = db.collection("projects").document(project).collection("files").document(filename).set(docData);
+    try {
+      System.out.println("File " + filename + " uploaded at time " + future.get().getUpdateTime());
+    } catch (Exception e) {
+      System.out.println(e);
+    }
+  }
+
+  /** 
+   * Returns a URL that points to the uploaded file, or null if the user didn't upload a file. 
+   */
+  @Nullable
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
```

Comment:
This is tying your back end to your front end. It's ok to say that your service only handles one file upload at a time. You might even consider throwing an exception if you get more than 1.

Reply:
Changed so that the backend support multiple file uploads